commands/general/help.js

```diff
@@ -1,55 +1,95 @@
-      usage: "help [command]",
-      category: "System",
+      description: "Get help on a command, command category, or a setting",
       extended: "This command will display all available commands for your permission level, with the additonal option of getting per command information when you run 'help <command name>'.",
-      hidden: true,
-      aliases: ["h", "halp", "commands"]
+      category: "System",
+      usage: "help <category/command/setting> [page-num]"
     });
   }
 
-  async run(message, args, level) {
-    const settings = message.settings;
-    if (!args[0]) {
-      try {
-        const myCommands = message.guild ? this.client.commands.filter(cmd => this.client.levelCache[cmd.conf.permLevel] <= level && cmd.conf.hidden !== true) : this.client.commands.filter(cmd => this.client.levelCache[cmd.conf.permLevel] <= level && cmd.conf.hidden !== true && cmd.conf.guildOnly !== true);
-        const commandNames = myCommands.keyArray();
-        const longest = commandNames.reduce((long, str) => Math.max(long, str.length), 0);
-        let currentCategory = "";
-        let output = `= Command List =\n\n[Use ${settings.prefix}help <commandname> in a guild channel for details]\n`;
-        const sorted = myCommands.array().sort((p, c) => p.help.category > c.help.category ? 1 :  p.help.name > c.help.name && p.help.category === c.help.category ? 1 : -1 );
-        sorted.forEach( c => {
-          const cat = c.help.category.toProperCase();
-          if (currentCategory !== cat) {
-            output += `\u200b\n== ${cat} ==\n`;
-            currentCategory = cat;
+  async run(message, [type, page], level) {
+
+    const embed = new MessageEmbed()
+      .setTimestamp()
+      .setColor(message.guild.me.roles.highest.color || 0x00AE86)
+      .setFooter(`Requested by ${message.author.tag}`, message.author.avatarURL()); 
+    // Preload MessageEmbed.
+  
+    // Here we sort out categories in case the user did not provide an argument.
+    
+    let currentCategory = "";
+    const sorted = this.client.commands.sort((p, c) => p.help.category > c.help.category ? 1 :  p.help.name > c.help.name && p.help.category === c.help.category ? 1 : -1 );
+    if (!type) {
+      const description = `Command category list\n\nUse \`${message.settings.prefix}help <category>\` to find commands for a specific category`;
+      const output = sorted.filter(c => !(level < 10 && c.help.category == "Owner") || !(c.help.category === "NSFW" && !message.channel.nsfw)).map(c => {
+        const cat = c.help.category.toProperCase();
+        if (currentCategory !== cat && !type) {
+          currentCategory = cat;
+          return `\n\`${message.settings.prefix}help ${cat.toLowerCase()}\` | Shows ${cat} commands`;
+        }
+      }).join("");
+
+      embed.setDescription(description)
+        .addField("Categories", output);
+
+      return message.channel.send({embed});
+    } else {
+      let lol = 0;
+      sorted.forEach(c => {
```

Comment:
forEach is asyncrous means that its not blocking and this code could cause a race condition